### PR TITLE
[FEATURE] Lora support in 2.3

### PR DESCRIPTION
**Pull request by lstein**
NOTE: This PR works with `diffusers` models **only**. As a result InvokeAI is now converting all legacy checkpoint/safetensors files into diffusers models on the fly. This introduces a bit of extra delay when loading legacy models. You can avoid this by converting the files to diffusers either at import time, or after the fact.

# Instructions:

1. Download LoRA .safetensors files of your choice and place in  `INVOKEAIROOT/loras`. Unlike the draft version of this PR, the file names can now contain underscores and hyphens. Names with arbitrary unicode characters are not supported.

2. Add `withLora(lora-file-basename,weight)` to your prompt. The  weight is optional and will default to 1.0. A few examples, assuming that a LoRA file named `loras/sushi.safetensors` is present:

```
family sitting at dinner table eating sushi withLora(sushi,0.9)
family sitting at dinner table eating sushi withLora(sushi, 0.75)
family sitting at dinner table eating sushi withLora(sushi)
```

Multiple `withLora()` prompt fragments are allowed. The weight can be arbitrarily large, but the useful range is roughly 0.5 to 1.0. Higher weights make the LoRA's influence stronger. The last version of the syntax, which uses the default weight of 1.0, is waiting on the next version of the Compel library to be released and may not work at this time.

In my limited testing, I found it useful to reduce the CFG to avoid oversharpening. Also I got better results when running the LoRA on top of the model on which it was based during training.

Don't try to load a SD 1.x-trained LoRA into a SD 2.x model, and vice versa. You will get a nasty stack trace. This needs to be cleaned up.

3. You can change the location of the `loras` directory by passing the  `--lora_directory` option to `invokeai.

Documentation can be found in docs/features/LORAS.md.

Note that this PR incorporates the unmerged 2.3.3 PR code (#3058) and bumps the version number up to 2.3.4a0.

A zillion thanks to @felorhik, @neecapp and many others for this implementation. @blessedcoolant and I just did a little tidying up.

### CODE REVIEW

**Comment by blessedcoolant** [comment]
Working as advertised.

I redid the prompt configuration to handle pretty much every case / scenario / filename / typos. Please have a look and update if you think its good.

```python
def configure_prompt(self, prompt: str) -> str:
        self.clear_loras()

        lora_match_pattern = re.compile(r"withLora\(([^\)]+)\)")
        lora_matches = re.findall(lora_match_pattern, prompt)

        for match in lora_matches:
            if ',' in match:
                name = match.split(',')[0].strip()
                try:
                    mult = float(match.split(',')[1].strip())
                except Exception:
                    mult = 1.0
            else:
                name = match.strip()
                mult = 1.0
            
            self.set_lora(name, mult)
        
        # remove lora and return prompt to avoid the lora prompt causing issues in inference
        return re.sub(lora_match_pattern, "", prompt)
```

**Review by blessedcoolant — APPROVED** [review]

**Comment by blessedcoolant** [comment]
Might also need to add a global arg to point to a custom directory for the lora models.

**Comment by blessedcoolant** [comment]
Gave it another try.

- I receive multiple `UserWarning: TypedStorage is deprecated.` on first run. This could be a Pytorch 2.0 thing. But probably might still be a good ideas to suppress it. 
- The `configure_prompt` regex fails in quite a few cases.
-- When the prompt is not a 100% match because of spacing issues. For example: `withLora(model_name ,weight)` is not recognized because of the space.
-- When an alphabet is passed to the weight, it fails.  Rather than reverting to 1.0

For the `configure_prompt`, I still recommend the function above. It'd be a hassle to try and regex match what is typed in between the brackets of `withLora`. Instead I think we're better of getting everything between the brackets and then parsing that. Because when a user types something in the brackets of `withLora`, we presume 100% that they're trying to invoke a lora. And once this is established, I think we just check for `,` in the string (which is our designated syntax) ---
-- If `,` is found, we split at the first instance and pass the first half to the `name` and the second to `mult`. In this case, we add an exception to `mult` to default to 1.0 when it fails on any case of `float` conversion. This will act as a fail safe for any half entries like `1.` or alphabetical entries.
- If no `,` found in string, we just apply the name and default the `mult` to 1.0.

I think that covers all possible scenarios that could occur. But I'll let you decide on it.

**Comment by damian0815** [comment]
why is the `configure_prompt` regex being used when compel already parses `withLora()` successfully?

**Comment by blessedcoolant** [comment]
Add Lora to WebUI.

**Comment by lstein** [comment]
I just wasn't sure where to fit Compel in. Do I make a new call to Compel.parse_prompt_string() in the `configure_prompt()` method of KohyaLoraManager? 

At this point Compel has already been used to get the conditioning, so presumably it's already been used to parse the prompt.

**Comment by lstein** [comment]
> I receive multiple UserWarning: TypedStorage is deprecated. on first run. This could be a Pytorch 2.0 thing. But probably might still be a good idea to suppress it.

I've been seeing these too whenever a `.safetensors` file is loaded. It appears to be a safetensors.torch issue and I'm assuming that a safetensors update is coming to fix it. I'm uncomfortable just suppressing all deprecation warnings.

**Comment by blessedcoolant** [comment]
> I've been seeing these too whenever a `.safetensors` file is loaded. It appears to be a safetensors.torch issue and I'm assuming that a safetensors update is coming to fix it. I'm uncomfortable just suppressing all deprecation warnings.

No worries. Can leave it on for now. Not a huge issue.

**Comment by lstein** [comment]
Oh yea

> Lora

Great work! But if you're going to expose the list of LoRAs to the web UI, then could you do something similar with the textual inversion triggers? You could even expose the list of HuggingFace concepts, which is cached on the back end.

**Comment by lstein** [comment]
This PR also breaks direct running of legacy checkpoints, in case you hadn't noticed. I plan to deal with this by ripping out all the legacy generation code in the same way as in `main`. Will be fun.

**Comment by blessedcoolant** [comment]
> Great work! But if you're going to expose the list of LoRAs to the web UI, then could you do something similar with the textual inversion triggers? You could even expose the list of HuggingFace concepts, which is cached on the back end.

Think we're really trying to avoid more sockets or changes to the web server while @psychedelicious works on the port. Added this one coz it felt weird to have a Lora release without it showing anywhere in the UI. People might not even notice.

If its not a big worry, I could expose TI's in the same way too maybe. But its just adding buttons and not really an intuitive UI. The best way to do this would be with autocomplete (which I don't think chakra has) and we'll need to build from scratch.

I'll take a look and see if there's a quick solution for this. If not, we'll do it up in 3.0.

**Comment by blessedcoolant** [comment]
> This PR also breaks direct running of legacy checkpoints, in case you hadn't noticed. I plan to deal with this by ripping out all the legacy generation code in the same way as in `main`. Will be fun.

Oh boy. Did not test old checkpoints. I haven't bothered with them since we move to `diffusers`.

**Comment by damian0815** [comment]
compel latest main fixes the issue with default weight=1 not being correctly parsed - haven't pushed to pypi yet though.

re: the `configure_prompt` stuff. as i wrote on Discord there's some issues here, namely that the kohya lora files are being loaded at the wrong time (far too early) and are not guaranteed to be cleaned up safely. instead, the lora load/unload should all go inside the `custom_attention_context()` `ContextManager` on the `InvokeAIDiffuserComponent`.

possibly, this can be achieved by limiting the lifetime of the `KohyaLoraManager` to completely inside this context - it would get constructed when `custom_attention_context()` was called and would then be cleaned up and destroyed when the context exits. this would also mean that the `lora_manager` instance on `StableDiffusionGeneratorPipeline` can go away (like it should). 

it should also be possible to subclass `LoraCondition` into `LoraCondition_Diffusers` and `LoraCondition_Kohya`, but this might involve breaking the `KohyaLoraManager` class up and putting parts of its implementation in there and parts in a new `KohyaLoraContext` object. 

i guess the same should be done with the PeftManager.

**Comment by damian0815** [comment]
@lstein lmk how you want to proceed with the above. i can possibly find some time to re-engineer this, which might be better than trying to explain to @jram why it's not suitable the way it is and what needs to be done to fix it.

File: ldm/modules/kohya_lora_manager.py
Comment (StAlKeR7779) [inline ldm/modules/kohya_lora_manager.py:26]: Alpha can be None or 0, that means unscalled layer.
Source:
https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py#L21
https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py#L51

File: ldm/modules/kohya_lora_manager.py
Comment (StAlKeR7779) [inline ldm/modules/kohya_lora_manager.py:309]: This regex missing lora's with numbers in name.

File: invokeai/backend/invoke_ai_web_server.py
Comment (StAlKeR7779) [inline invokeai/backend/invoke_ai_web_server.py:493]: Found some loras on civitai with `.pt` extension instead of `.ckpt`.
Maybe scan files with this extension too?

File: ldm/modules/kohya_lora_manager.py
Comment (StAlKeR7779) [inline ldm/modules/kohya_lora_manager.py:321]: Same as above - maybe scan for `pt` extension too?

Comment (lstein) [reply ldm/modules/kohya_lora_manager.py]: Fixed. Thanks for the suggestion.

File: invokeai/backend/invoke_ai_web_server.py
Comment (lstein) [reply invokeai/backend/invoke_ai_web_server.py:493]: Fixed. Thanks for the suggestion.

File: ldm/modules/kohya_lora_manager.py
Comment (lstein) [reply ldm/modules/kohya_lora_manager.py:29]: I'll take your word for it. Implemented.

File: ldm/modules/kohya_lora_manager.py
Comment (lstein) [reply ldm/modules/kohya_lora_manager.py]: Superseded. lora parsing now part of Compel.

**Comment by lstein** [comment]
> compel latest main fixes the issue with default weight=1 not being correctly parsed - haven't pushed to pypi yet though.
> 
> re: the `configure_prompt` stuff. as i wrote on Discord there's some issues here, namely that the kohya lora files are being loaded at the wrong time (far too early) and are not guaranteed to be cleaned up safely. instead, the lora load/unload should all go inside the `custom_attention_context()` `ContextManager` on the `InvokeAIDiffuserComponent`.
> 
> possibly, this can be achieved by limiting the lifetime of the `KohyaLoraManager` to completely inside this context - it would get constructed when `custom_attention_context()` was called and would then be cleaned up and destroyed when the context exits. this would also mean that the `lora_manager` instance on `StableDiffusionGeneratorPipeline` can go away (like it should).
> 
> it should also be possible to subclass `LoraCondition` into `LoraCondition_Diffusers` and `LoraCondition_Kohya`, but this might involve breaking the `KohyaLoraManager` class up and putting parts of its implementation in there and parts in a new `KohyaLoraContext` object.
> 
> i guess the same should be done with the PeftManager.

I created a KohyaLoraContext object that I apply just before image generation starts. It seems to work properly to apply the loaded lora to the model before generation and remove it from the model after generation.

I think I understand what you are recommending here, but if it involves creating a new KohyaLoraManager and tearing it down each time, I'm not sure that this is what we want. I would like to keep the loaded Kohya models in memory so that once they are loaded they can be temporarily applied to the models when the prompt requests them without loading from disk again. Of course they aren't very large so this may not make a performance difference one way or another.

File: ldm/modules/kohya_lora_manager.py
Comment (neecapp) [inline ldm/modules/kohya_lora_manager.py:344]:
```suggestion
```

File: ldm/modules/lora_manager.py
Comment (neecapp) [inline ldm/modules/lora_manager.py:22]:
```suggestion
    def __init__(self, name, weight: float = 1.0, kohya_manager: Optional[KohyaLoraManager]=None):
```

File: ldm/modules/lora_manager.py
Comment (neecapp) [inline ldm/modules/lora_manager.py:43]: Maybe something like this would simplify adding future types (such as LyCORIS), omitted all logging/etc.:

```python
# using Any here, but should be type of `model`
Stage = Callable[[str, float, Any], bool]

class LoraConditionChain:
    def __init__(self, stages: list[Stage]):
        self.stages = set(stages)

    def make_chain(self, name: str, weight: Optional[float]):
        stages = self.stages

        def call(model):
            for s in stages:
                if s(name, weight or 1.0, model):
                    return
            print(f"   ** Failed to load LoRA {name}")

        return call

class KohyaLoraCondition:
    def __init__(self, manager: KohyaLoraManager):
        self.manager = manager

    def __call__(self, name: str, weight: float, model):
        return self.manager.apply_lora_model(name, weight)  # would require updating KohyaLoraManager.apply_lora_model to return True/False depending on if the lora was found and loaded


class DiffuserLoraCondition:
    def __call__(self, name: str, weight: float, model):
        path = Path(global_lora_models_dir(), name)
        if path.is_dir():
            if model.load_attn_procs:
                file = Path(path, "pytorch_lora_weights.bin")
                if file.is_file():
                    print(f">> Loading LoRA: {path}")
                    model.load_attn_procs(path.absolute().as_posix())
                    return True
        return False
```

Then, inside of `LoraManager`, you can simply create a chain:

```python
class LoraManager:
    def __init__(self, pipe):
        # Kohya class handles lora not generated through diffusers
        self.kohya = KohyaLoraManager(pipe, global_lora_models_dir())
        self.lora_condition = LoraConditionChain([KohyaCondition(self.kohya), DiffuserCondition()])

    def set_loras_conditions(self, lora_weights: list):
        conditions = []
        if len(lora_weights) > 0:
            for lora in lora_weights:
                conditions.append(self.lora_condition.make_chain(lora.model, lora.weight))

        if len(conditions) > 0:
            return conditions

        return None

    def kohya_context(self) -> KohyaLoraContext:
        return KohyaLoraContext(self.kohya)
```

This is also missing Peft, but... unsure that even matters at this point, there are no Peft LoRA in the wild that I have ever seen.

**Review by neecapp — COMMENTED** [review]
Thank you for picking this up. I haven't had much free time recently to work on helping create a proper implementation.

File: invokeai/frontend/src/features/parameters/components/LoraManager/LoraManager.tsx
Comment (lstein) [inline invokeai/frontend/src/features/parameters/components/LoraManager/LoraManager.tsx:44]: Loading the webUI on chrome is causing high CPU load on the InvokeAI tab. Looking at the performance profile, I find this function being called repeatedly many times per second.

File: ldm/modules/kohya_lora_manager.py
Comment (lstein) [reply ldm/modules/kohya_lora_manager.py:259]: It is used now during cleanup.

**Comment by lstein** [comment]
@blessedcoolant I have a private branch in which I added a WebUI button for viewing Textual Inversion triggers, modeled after the one you wrote for Loras. Unfortunately it has the same apparent problem of causing high load on the browser. I'd love to add this in after the load problem is tracked down.

File: ldm/modules/lora_manager.py
Comment (blessedcoolant) [inline ldm/modules/lora_manager.py:12]: Optional is not imported here.

File: invokeai/frontend/src/features/parameters/components/LoraManager/LoraManager.tsx
Comment (blessedcoolant) [reply invokeai/frontend/src/features/parameters/components/LoraManager/LoraManager.tsx:44]: Looking into it.

Comment (neecapp) [reply invokeai/frontend/src/features/parameters/components/LoraManager/LoraManager.tsx:44]: Pretty sure your issue is from the `useEffect` below on [line 70](https://github.com/invoke-ai/InvokeAI/pull/3072/files/8518f8c2acdaf59ac1948999f0130e725b041548#diff-dae5156119f358589060f8ed64a6ff8ea06f884c4aa8dc1af01f0b5c98b2df84R70). You are listening to `loraItems`, which are being set within the effect. You are making a new array and the effect is checking on strict equality, which means it is never equal.

Changing to `  }, [foundLoras, handleLora, renderLoraOption]);` resolves the issue for me.

File: ldm/modules/lora_manager.py
Comment (lstein) [reply ldm/modules/lora_manager.py]: This looks promising. I'll take a closer look tomorrow.

For what it's worth, I've only tested this so far with Kohya LoRAs. I have no idea if the Peft and diffusers LoRAs are working.

**Comment by StAlKeR7779** [comment]
> @blessedcoolant I have a private branch in which I added a WebUI button for viewing Textual Inversion triggers, modeled after the one you wrote for Loras. Unfortunately it has the same apparent problem of causing high load on the browser. I'd love to add this in after the load problem is tracked down.

I don't think that this button works fine with TIs, because unlike to loras, TIs can be both in positive and negative prompts and there are to many of them to properly show list.
For helping user with TIs I think autocompletetion fits better. With autocompletetion we can show only 3-5 most sutable options.

**Comment by blessedcoolant** [comment]
> I don't think that this button works fine with TIs, because unlike to loras, TIs can be both in positive and negative prompts and there are to many of them to properly show list. For helping user with TIs I think autocompletetion fits better. With autocompletetion here can be shown only 3-5 most suitable options.

I tried implementing the same kind for TI's on my end. Not feeling too great. Basically need to scan every word because a TI can be named anything. I fixed up the TI list to be a short dropdown but I still think all of these need a proper autocomplete component that we do not have at the moment.

And that is out of the scope of this PR. Right now, we get a basic implementation in so the LoRA's are working and then we refine the experience in 3.0 imo.

**Comment by lstein** [comment]
So consensus is to take the TI button out completely or just limit it to the trigger terms in local embed files? I think that the overload bug is still there in the TI button.

**Comment by blessedcoolant** [comment]
> So consensus is to take the TI button out completely or just limit it to the trigger terms in local embed files? I think that the overload bug is still there in the TI button.

I have a version on my end that is refactored. The thing we need to decide is if we want to keep the prompt typing to interact with TI words. Coz effectively any word can be TI word and there's no specific regex, we'll be running constant checks which can be really taxing on the browser.

So I think we keep the button functionality to add and remove the keywords but not have any highlighting or manually typed support like the Loras do.

File: invokeai/frontend/src/features/parameters/store/generationSlice.ts
Comment (neecapp) [inline invokeai/frontend/src/features/parameters/store/generationSlice.ts:90]: This should really be split out of `setPrompt` and debounced. It's not obvious, but this will cause input lag for prompts when there is a large number of LoRA.

Comment (blessedcoolant) [reply invokeai/frontend/src/features/parameters/store/generationSlice.ts:90]: I agree. I'm a tad bit busy but I'll to get to it. Maybe @psychedelicious can have a look if he's free too.

**Comment by lstein** [comment]
> > So consensus is to take the TI button out completely or just limit it to the trigger terms in local embed files? I think that the overload bug is still there in the TI button.
> 
> > I have a version on my end that is refactored. The thing we need to decide is if we want to keep the prompt typing to interact with TI words. Coz effectively any word can be TI word and there's no specific regex, we'll be running constant checks which can be really taxing on the browser.
> 
> > So I think we keep the button functionality to add and remove the keywords but not have any highlighting or manually typed support like the Loras do.
> 
> I agree. No need to keep trying to find the TI trigger in the prompt. A simple inserter will work fine.
> 
> > Also for the TI drop down, I think user TI's should show up first followed by hugging face concepts lib. I could sort this on the frontend but I think it is better handled by Python.
> 
> This is already what it does. The python backend adds the user's TIs first, followed by an alpha sorted list of all hugging face concepts with at least 6 likes.

Once again I inadvertently chose "edit" 'instead of "reply" and added my comments inline with blessedcoolant's.

**Comment by lstein** [comment]
@blessedcoolant Looks great! I like the option to turn on and off the huggingface concepts. Great compromise.

**Comment by blessedcoolant** [comment]
@lstein Check out my current implementation. It's already live on this PR. The setting is in the Settings menu. This doesn't disable HuggingFace Concepts. Just their display in the menu. If a user manually types it in, they'll still work and UI will reflect the embedding being used in the counter.

**Comment by lstein** [comment]
Like I said in my comment, I've tested it and like it very much. I'm
proposing that we change the configuration option label to "Show Textual
Inversions from HF Concepts Library" so that it matches up with the button
label "Add Textual Inversion".

Lincoln

**Comment by lstein** [comment]
I've committed the label change. It doesn't look untowardly long.

**Comment by lstein** [comment]
@blessedcoolant I think we're ready for a final review and merge. What do you think?

**Comment by blessedcoolant** [comment]
@lstein Yep. I think we can get one or two more people to test out the PR and we merge it. Seems good.

ps. I don't see the label change yet. I'm guessing you haven't pushed it still.

**Comment by lstein** [comment]
> @lstein Yep. I think we can get one or two more people to test out the PR and we merge it. Seems good.
> 
> ps. I don't see the label change yet. I'm guessing you haven't pushed it still.

Pushed now.

**Comment by blessedcoolant** [comment]
@lstein Found a bug.

The `.last_model` does not update when we run the app in gui mode with `--gui` ... Maybe just set it on model change rather than when the app quits? Because that effectively serves the same purpose and it won't interfere with app crashes or unexpected terminations?

**Review by ebr — COMMENTED** [review]
LGTM; only question is whether we're intentionally merging Nodes here.

**Comment by blessedcoolant** [comment]
@lstein This PR also needs compel to be bumped up.

File: pyproject.toml
Comment (ebr) [inline pyproject.toml:44]: a `nodes` dependency
```suggestion
```

**Review by ebr — APPROVED** [review]
approved from install/packaging perspective

File: pyproject.toml
Comment (lstein) [reply pyproject.toml:44]: The nodes shouldn't have been in there and have been removed (3.0 feature, not 2.3) so I don't think the fastapi changes are needed?

File: ldm/invoke/args.py
Comment (lstein) [reply ldm/invoke/args.py:526]: We are converting legacy .ckpt in memory on every load. If the user wishes to avoid this, they can do a conversion either in the CLI or or the WebUI. It is pretty transparent, except that the .ckpt models load more slowly than they used to.

File: ldm/invoke/concepts_lib.py
Comment (lstein) [reply ldm/invoke/concepts_lib.py:30]: That's along the right lines. To really do it right, we need to do some scaling to avoid a bias against more recent models that haven't had enough time to accumulate likes. Right now I'm happy to see how often we get complaints that we missed a good model that should have been on the list.

In 3.0, the WebUI dropdowns will disappear and we'll be able to autocomplete. The like threshold can go away at that point.

**Comment by damian0815** [comment]
I can't find anywhere the `LoraModuleWrapper`'s `clear_hooks()` method is called - this seems to be not ok, is this ok?

**Comment by lstein** [comment]
> I can't find anywhere the `LoraModuleWrapper`'s `clear_hooks()` method is called - this seems to be not ok, is this ok?

Really good question! The hooks are inserted into the diffusers pipeline at pipeline load time via the KohyaLoraManager's __init__ routine and never removed. The hooks remain attached to the pipeline as it is moved in and out of VRAM. In the event that the pipeline is purged from CPU RAM cache, all the LoRA objects are deleted and ultimately garbage collected. 

Tracing the code, it doesn't seem to me that there is a good reason to clear the hooks. If no LoRA condition is present, then the code that the hooks point to are no-ops. Clearing the hooks after a generation is completed looks easy to do, but reinitializing them when a LoRA condition is requested seems dicey as there are several layers of code to go through.

My feeling is that it is safe to keep the hooks attached. However, I will do some generations with and without the LoRA manager installed to ensure that the presence of the hooks don't themselves affect the generated image, memory consumption, or performance.

File: ldm/modules/kohya_lora_manager.py
Comment (StAlKeR7779) [inline ldm/modules/kohya_lora_manager.py:359]: This function is unused.
Also for me such definition controversial to `unload_applied_lora(lora_name)` definition.

Comment (lstein) [reply ldm/modules/kohya_lora_manager.py:359]: Correct. We don't call this. I expect it can be removed as it is probably dangerous to call.

Comment (lstein) [reply ldm/modules/kohya_lora_manager.py]: Both issues addressed in recent commit.

**Comment by Void2258** [comment]
Found a problem with the `.blend` function. It's trying to work with loras even without loras present.

`("girl","cat").blend(1,1)` (simplest possible test for blending) gives
`Traceback (most recent call last):
  File "F:\Invoke_AI\.venv\lib\site-packages\invokeai\backend\invoke_ai_web_server.py", line 1353, in generate_images
    self.generate.prompt2image(
  File "F:\Invoke_AI\.venv\lib\site-packages\ldm\generate.py", line 524, in prompt2image
    uc, c, extra_conditioning_info = get_uc_and_c_and_ec(
  File "F:\Invoke_AI\.venv\lib\site-packages\ldm\invoke\conditioning.py", line 67, in get_uc_and_c_and_ec
    should_use_lora_manager = model.peft_manager.should_use(positive_prompt.lora_weights)
AttributeError: 'Blend' object has no attribute 'lora_weights'`

File: ldm/invoke/conditioning.py
Comment (neecapp) [inline ldm/invoke/conditioning.py:69]: Unsure how beneficial this is, has anyone even tested peft formatted LoRA?

File: ldm/invoke/conditioning.py
Comment (neecapp) [inline ldm/invoke/conditioning.py:71]:
```suggestion
        if hasattr(positive_prompt, 'lora_weights') and model.lora_manager and should_use_lora_manager:
            lora_conditions = model.lora_manager.set_loras_conditions(positive_prompt.lora_weights)
```

**Comment by neecapp** [comment]
> Found a problem with the `.blend` function. It's trying to work with loras even without loras present.
> 
> `("girl","cat").blend(1,1)` (simplest possible test for blending) gives `Traceback (most recent call last):` ` File "F:\Invoke_AI\.venv\lib\site-packages\invokeai\backend\invoke_ai_web_server.py", line 1353, in generate_images self.generate.prompt2image(` ` File "F:\Invoke_AI\.venv\lib\site-packages\ldm\generate.py", line 524, in prompt2image uc, c, extra_conditioning_info = get_uc_and_c_and_ec(` ` File "F:\Invoke_AI\.venv\lib\site-packages\ldm\invoke\conditioning.py", line 67, in get_uc_and_c_and_ec should_use_lora_manager = model.peft_manager.should_use(positive_prompt.lora_weights)` `AttributeError: 'Blend' object has no attribute 'lora_weights'`

https://github.com/invoke-ai/InvokeAI/pull/3072/files#r1160306518 or similar should resolve that issue.

File: ldm/invoke/conditioning.py
Comment (lstein) [reply ldm/invoke/conditioning.py]: Indeed not. Where do you even find a peft formatted LoRA to test with?

File: ldm/invoke/conditioning.py
Comment (lstein) [reply ldm/invoke/conditioning.py]: Thanks. I solved it in a slightly different way, but same idea.

**Review by StAlKeR7779 — APPROVED** [review]

File: ldm/invoke/conditioning.py
Comment (neecapp) [reply ldm/invoke/conditioning.py]: I've never heard of anyone training them outside of random tests. Unfortunately, the example linked from huggingface/peft [is disabled](https://huggingface.co/spaces/smangrul/peft-lora-sd-dreambooth), and I cannot find any SD PEFT LoRA anywhere.

**Comment by damian0815** [comment]
> > Found a problem with the `.blend` function. It's trying to work with loras even without loras present.
> > `("girl","cat").blend(1,1)` (simplest possible test for blending) gives `Traceback (most recent call last):` ` File "F:\Invoke_AI\.venv\lib\site-packages\invokeai\backend\invoke_ai_web_server.py", line 1353, in generate_images self.generate.prompt2image(` ` File "F:\Invoke_AI\.venv\lib\site-packages\ldm\generate.py", line 524, in prompt2image uc, c, extra_conditioning_info = get_uc_and_c_and_ec(` ` File "F:\Invoke_AI\.venv\lib\site-packages\ldm\invoke\conditioning.py", line 67, in get_uc_and_c_and_ec should_use_lora_manager = model.peft_manager.should_use(positive_prompt.lora_weights)` `AttributeError: 'Blend' object has no attribute 'lora_weights'`
> 
> https://github.com/invoke-ai/InvokeAI/pull/3072/files#r1160306518 or similar should resolve that issue.

this is superseded with my latest commit + new compel version

**Comment by blessedcoolant** [comment]
I ran some tests. Seems to be good. If one or more of you can run some final tests, I think we need to get another party to approve this PR. Coz I don't think @lstein or @damian0815 or my own approval counts now coz we all contributed to this.

We merge this up and fix any pending bugs in another PR. If it's clean, it's good for a release.

**Pull request merged**
